compiler/luci-interpreter/src/kernels/Quantize.cpp

```diff
@@ -0,0 +1,159 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kernels/Quantize.h"
+#include "kernels/Utils.h"
+#include "PALQuantize.h"
+
+namespace luci_interpreter
+{
+namespace kernels
+{
+
+namespace
+{
+
+template <typename input_dtype> void call_requantize(const Tensor *input, Tensor *output)
+{
+  int32_t multiplier;
+  int32_t shift;
+
+  const double effective_output_scale = input->scale() / output->scale();
+  quantizeMultiplier(effective_output_scale, &multiplier, &shift);
+
+  const auto input_shape = getTensorShape(input);
+  const auto output_shape = getTensorShape(output);
+  const auto size = tflite::MatchingFlatSize(input_shape, output_shape);
+
+  const auto input_data = getTensorData<input_dtype>(input);
+
+  switch (output->element_type())
+  {
+    case loco::DataType::S8:
+      luci_interpreter_pal::Requantize(input_data, size, multiplier, shift, input->zero_point(),
+                                       output->zero_point(), getTensorData<int8_t>(output));
+      break;
+    case loco::DataType::U8:
+      luci_interpreter_pal::Requantize(input_data, size, multiplier, shift, input->zero_point(),
+                                       output->zero_point(), getTensorData<uint8_t>(output));
+      break;
+    case loco::DataType::S16:
+      luci_interpreter_pal::Requantize(input_data, size, multiplier, shift, input->zero_point(),
+                                       output->zero_point(), getTensorData<int16_t>(output));
+      break;
+    default:
+      throw std::runtime_error("Unsupported quantized type, yet!");
+  }
+}
+
+} // namespace
+
+Quantize::Quantize(const Tensor *input, Tensor *output) : Kernel({input}, {output}) {}
+
+void Quantize::configure()
+{
+  if (input()->element_type() == loco::DataType::FLOAT32)
+  {
+    LUCI_INTERPRETER_CHECK(output()->element_type() == loco::DataType::U8 ||
+                           output()->element_type() == loco::DataType::S8 ||
+                           output()->element_type() == loco::DataType::S16);
+  }
+  else if (input()->element_type() == loco::DataType::S16)
+  {
+    LUCI_INTERPRETER_CHECK(output()->element_type() == loco::DataType::S8 ||
+                           output()->element_type() == loco::DataType::S16);
+    if (output()->element_type() == loco::DataType::S16)
+    {
+      LUCI_INTERPRETER_CHECK(input()->zero_point() == 0 && output()->zero_point() == 0);
+    }
+  }
+  else if (input()->element_type() == loco::DataType::S8 ||
+           input()->element_type() == loco::DataType::U8)
+  {
+    // TODO: support S8 (U8) -> S16
+    LUCI_INTERPRETER_CHECK(output()->element_type() == loco::DataType::S8 ||
+                           output()->element_type() == loco::DataType::U8);
+  }
```

Comment:
Why `// TODO: support S8 (U8) -> S16`? As far as I understand we are support this requantization for now with `call_requantize` helper. IIUC we need just to extend `LUCI_INTERPRETER_CHECK`'` condition.

Reply:
Also, how about switch-case here for `input->element_type()`?

Reply:
> Why `// TODO: support S8 (U8) -> S16`? As far as I understand we are support this requantization for now with `call_requantize` helper. IIUC we need just to extend `LUCI_INTERPRETER_CHECK`'` condition.

I am not sure Quantize kernel can work properly with such quantization due to tflite Quantize kernel doesn't support such quantization:   https://github.com/tensorflow/tensorflow/blob/acddd82daca0c2c3e374fb8ed9607a5771918c2e/tensorflow/lite/kernels/quantize.cc#L126
```c++
   // Requantize use case.
    if (input->type == kTfLiteInt16) {
      TF_LITE_ENSURE(context, output->type == kTfLiteInt8 ||
                                  output->type == kTfLiteInt16 ||
                                  output->type == kTfLiteInt32);
    } else if (input->type == kTfLiteInt32) {
      TF_LITE_ENSURE(
          context, output->type == kTfLiteInt8 || output->type == kTfLiteInt16);
    } else {
      TF_LITE_ENSURE(context,
                     input->type == kTfLiteInt8 || input->type == kTfLiteUInt8);
      TF_LITE_ENSURE(
          context, output->type == kTfLiteUInt8 || output->type == kTfLiteInt8);
    }
```

Reply:
But If you think It will be ok, I will add this quantization case.

Reply:
I checked `tflite::Requantize` impl and looks like it works correctly for any quantized datatype depending only on `dtype`'s min/max and quantization zero point:
https://github.com/tensorflow/tensorflow/blob/acddd82daca0c2c3e374fb8ed9607a5771918c2e/tensorflow/lite/kernels/internal/reference/requantize.h#L51-L62

Reply:
Done